File: hash/isomorphic_strings.cpp
//{ Driver Code Starts
// C++ program to check if two strings are isomorphic

#include <bits/stdc++.h>

using namespace std;
#define MAX_CHARS 256

// } Driver Code Ends
class Solution {
public:
  // Function to check if two strings are isomorphic.
  bool areIsomorphic(string str1, string str2) {
    // Your code here
    if (str1.size() != str2.size())
      return false;
    unordered_map<char, char> m;
    for (int i = 0; i < str1.size(); ++i) {
      if (m.find(str1[i]) == m.end()) {
        m[str1[i]] = str2[i];
      } else {
        if (m[str1[i]] != str2[i])
          return false;
      }
    }
    m.clear();
    for (int i = 0; i < str1.size(); ++i) {
      if (m.find(str2[i]) == m.end()) {
        m[str2[i]] = str1[i];
      } else {
        if (m[str2[i]] != str1[i])
          return false;
      }
    }
    return true;
  }
};

//{ Driver Code Starts.

// Driver program
int main() {
  int t;
  cin >> t;
  string s1, s2;
  while (t--) {
    cin >> s1;
    cin >> s2;
    Solution obj;
    cout << obj.areIsomorphic(s1, s2) << endl;
  }

  return 0;
}
// } Driver Code Ends

File: hash/isomorphic_strings.py
# User function Template for python3


class Solution:
    # Function to check if two strings are isomorphic.
    def areIsomorphic(self, str1, str2):
        d = {}
        d2 = {}
        n = len(str1)
        if n != len(str2):
            return False
        for i in range(n):
            if str1[i] in d:
                if d[str1[i]] != str2[i]:
                    return False
            else:
                if str2[i] in d2:
                    return False
                d2[str2[i]] = str1[i]
                d[str1[i]] = str2[i]

        return True


# {
# Driver Code Starts
# Initial Template for Python 3

import atexit
import io
import sys
from collections import defaultdict

_INPUT_LINES = sys.stdin.read().splitlines()
input = iter(_INPUT_LINES).__next__
_OUTPUT_BUFFER = io.StringIO()
sys.stdout = _OUTPUT_BUFFER


@atexit.register
def write():
    sys.__stdout__.write(_OUTPUT_BUFFER.getvalue())


if __name__ == "__main__":
    t = int(input())
    for i in range(t):
        s = str(input())
        p = str(input())
        ob = Solution()
        if ob.areIsomorphic(s, p):
            print(1)
        else:
            print(0)
# } Driver Code Ends
